### PR TITLE
Add a component style categorizer

### PR DESCRIPTION
**Pull request by blowery**
Adds a little utility that tries to score remaining component styles to migrate.

### CODE REVIEW

File: bin/categorize-component-styles.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ */
+function hasNonCompliantToplevelSelectors( f, name ) {
+	let topLevelSelectors;
+	const re = /^\.([\w_-]+)/gm;
+	while ( ( topLevelSelectors = re.exec( f ) ) !== null ) {
+		//console.log( topLevelSelectors[1] );
+		if ( ! topLevelSelectors[ 1 ].startsWith( name ) ) {
+			// suspect
+			//console.log( 'saw %s, expected %s', topLevelSelectors[1], name );
+			return {
```
Comment (blowery) [inline bin/categorize-component-styles.js]: maybe add these up and jump the score 🗡

**Comment by belcherj** [comment]
SCORE: 11
my-sites/stats/stats-module/style
   contains @import, uses @extend, uses @include, odd top level selector

SCORE: 7
blocks/support-article-dialog/style
   contains @import, uses @include

SCORE: 6
blocks/post-item/style
   uses @extend, uses @include, odd top level selector
me/notification-settings/blogs-settings/style
   uses @extend, uses @include, odd top level selector
my-sites/checkout/checkout/style
   uses @extend, uses @include, odd top level selector
my-sites/pages/style
   uses @extend, uses @include, odd top level selector

SCORE: 5
auth/style
   uses @include, odd top level selector
blocks/conversation-caterpillar/style
   uses @include, odd top level selector
blocks/conversations/style
   uses @include, odd top level selector
blocks/follow-button/style
   uses @include, odd top level selector
blocks/gdpr-banner/style
   uses @include, odd top level selector
blocks/google-my-business-stats-nudge/style
   uses @include, odd top level selector
blocks/image-selector/style
   uses @include, odd top level selector
blocks/inline-help/style
   uses @include, odd top level selector
blocks/login/style
   uses @include, odd top level selector
blocks/nps-survey/style
   uses @include, odd top level selector
blocks/privacy-policy-banner/style
   uses @include, odd top level selector
blocks/sharing-preview-pane/style
   uses @include, odd top level selector
blocks/site-address-changer/style
   uses @include, odd top level selector
blocks/site/style
   uses @include, odd top level selector
blocks/term-tree-selector/style
   contains @import
blocks/upload-image/style
   uses @include, odd top level selector
blocks/video-editor/style
   uses @include, odd top level selector
components/card/style
   uses @include, odd top level selector
components/date-picker/style
   uses @include, odd top level selector
components/dialog/style
   uses @include, odd top level selector
components/domains/domain-suggestion/style
   uses @include, odd top level selector
components/domains/featured-domain-suggestions/style
   contains @import
components/foldable-card/style
   uses @include, odd top level selector
components/happychat/style
   uses @include, odd top level selector
components/seo/style
   uses @include, odd top level selector
components/site-selector/style
   uses @include, odd top level selector
jetpack-connect/style
   uses @include, odd top level selector
layout/masterbar/crowdsignal
   uses @include, odd top level selector
layout/masterbar/oauth-client
   uses @include, odd top level selector
layout/masterbar/style
   uses @include, odd top level selector
layout/sidebar/style
   uses @include, odd top level selector
me/help/help-results/style
   uses @include, odd top level selector
me/help/help-search/style
   uses @include, odd top level selector
me/notification-settings/push-notification-settings/style
   uses @include, odd top level selector
me/notification-settings/settings-form/style
   uses @include, odd top level selector
me/purchases/purchases-site/style
   uses @include, odd top level selector
my-sites/activity/activity-log-banner/style
   uses @include, odd top level selector
my-sites/activity/activity-log-confirm-dialog/style
   uses @include, odd top level selector
my-sites/activity/activity-log-item/style
   uses @include, odd top level selector
my-sites/activity/activity-log/style
   uses @include, odd top level selector
my-sites/activity/filterbar/style
   uses @include, odd top level selector
my-sites/checkout/cart/style
   uses @include, odd top level selector
my-sites/checkout/checkout-thank-you/google-voucher/style
   uses @include, odd top level selector
my-sites/domains/domain-management/email/style
   uses @include, odd top level selector
my-sites/domains/domain-management/style
   uses @include, odd top level selector
my-sites/domains/domain-search/style
   uses @include, odd top level selector
my-sites/earn/ads/style
   uses @include, odd top level selector
my-sites/exporter/style
   uses @include, odd top level selector
my-sites/importer/style
   uses @include, odd top level selector
my-sites/media-library/upload-button
   uses @include, odd top level selector
my-sites/plan-features/style
   uses @include, odd top level selector
my-sites/sidebar-navigation/style
   uses @include, odd top level selector
my-sites/sidebar/style
   uses @include, odd top level selector
my-sites/stats/most-popular/style
   uses @include, odd top level selector
my-sites/stats/stats-site-overview/style
   uses @include, odd top level selector
my-sites/stats/style
   uses @include, odd top level selector
my-sites/theme/style
   uses @include, odd top level selector
my-sites/themes/style
   uses @include, odd top level selector
my-sites/themes/themes-magic-search-card/style
   uses @include, odd top level selector
my-sites/upgrade-nudge/style
   uses @include, odd top level selector

SCORE: 4
my-sites/domains/domain-management/transfer/transfer-out/style
   uses @extend, odd top level selector
my-sites/stats/geochart/style
   uses @extend, odd top level selector

SCORE: 3
blocks/comments/style
   uses @extend, uses @include
components/button/style
   odd top level selector
components/drop-zone/style
   odd top level selector
components/ellipsis-menu/style
   odd top level selector
components/forms/form-currency-input/style
   odd top level selector
components/happychat/agent-w/style
   odd top level selector
components/info-popover/style
   odd top level selector
components/input-chrono/style
   odd top level selector
components/marketing-survey/cancel-purchase-form/style
   odd top level selector
components/pagination/style
   uses @extend, uses @include
components/post-excerpt/style
   uses @extend, uses @include
components/segmented-control/style
   odd top level selector
layout/guided-tours/style
   odd top level selector
layout/nps-survey-notice/style
   odd top level selector
lib/abtest/test-helper/style
   odd top level selector
lib/keyboard-shortcuts/style
   odd top level selector
lib/preferences-helper/style
   odd top level selector
me/help/help-courses/style
   uses @extend, uses @include
me/notification-settings/comment-settings/style
   odd top level selector
me/notification-settings/wpcom-settings/style
   odd top level selector
me/sidebar-navigation/style
   odd top level selector
me/sidebar/style
   odd top level selector
my-sites/activity/activity-log-tasklist/style
   odd top level selector
my-sites/customize/style
   odd top level selector
my-sites/domains/domain-management/components/email-verification/style
   odd top level selector
my-sites/domains/domain-management/components/icann-verification/style
   odd top level selector
my-sites/pages/blog-posts-page/style
   uses @extend, uses @include
my-sites/posts/style
   odd top level selector
my-sites/stats/post-performance/style
   odd top level selector
my-sites/stats/stats-chart-tabs/style
   uses @extend, uses @include
my-sites/stats/stats-comments/style
   odd top level selector
my-sites/stats/stats-countries/style
   odd top level selector
my-sites/stats/stats-error/style
   odd top level selector

SCORE: 2
blocks/comment-button/style
   uses @include
blocks/conversation-follow-button/style
   uses @include
blocks/credit-card-form/style
   uses @include
blocks/domain-to-plan-nudge/style
   uses @include
blocks/edit-gravatar/style
   uses @include
blocks/eligibility-warnings/style
   uses @include
blocks/image-editor/style
   uses @include
blocks/like-button/style
   uses @include
blocks/payment-methods/style
   uses @include
blocks/post-edit-button/style
   uses @include
blocks/post-likes/style
   uses @include
blocks/product-purchase-features-list/style
   uses @include
blocks/stats-navigation/style
   uses @include
blocks/taxonomy-manager/style
   uses @include
blocks/term-form-dialog/style
   uses @include
blocks/upgrade-nudge-expanded/style
   uses @include
blocks/user-mentions/style
   uses @include
components/banner/style
   uses @include
components/card-heading/style
   uses @include
components/checklist/style
   uses @include
components/credit-card-form-fields/style
   uses @include
components/date-range/style
   uses @include
components/domains/contact-details-form-fields/style
   uses @include
components/domains/domain-product-price/style
   uses @include
components/domains/domain-search-results/style
   uses @include
components/domains/domain-transfer-suggestion/style
   uses @include
components/domains/example-domain-suggestions/style
   uses @include
components/domains/map-domain-step/style
   uses @include
components/domains/search-filters/style
   uses @include
components/domains/transfer-domain-step/style
   uses @include
components/domains/use-your-domain-step/style
   uses @include
components/email-verification/email-verification-dialog/style
   uses @include
components/empty-content/style
   uses @include
components/environment-badge/style
   uses @include
components/faq/style
   uses @include
components/formatted-header/style
   uses @include
components/forms/form-buttons-bar/style
   uses @include
components/forms/form-range/style
   uses @include
components/forms/form-text-input-with-affixes/style
   uses @include
components/global-notices/style
   uses @include
components/happiness-support/style
   uses @include
components/header-cake/style
   uses @include
components/keyed-suggestions/style
   uses @include
components/language-picker/style
   uses @include
components/legend-item/style
   uses @include
components/logged-out-form/style
   uses @include
components/mobile-back-to-sidebar/style
   uses @include
components/notice/style
   uses @include
components/popover/style
   uses @include
components/purchase-detail/style
   uses @include
components/rewind-credentials-form/style
   uses @include
components/section-nav/style
   uses @include
components/site-title-example/style
   uses @include
components/sub-masterbar-nav/style
   uses @include
components/theme/style
   uses @include
components/upgrades/gsuite/gsuite-dialog/style
   uses @include
components/web-preview/style
   uses @include
components/wizard/style
   uses @include
jetpack-connect/jetpack-new-site/style
   uses @include
layout/style
   uses @include
mailing-lists/style
   uses @include
me/account-close/style
   uses @include
me/account-password/style
   uses @include
me/billing-history/style
   uses @include
me/get-apps/style
   uses @include
me/happychat/style
   uses @include
me/help/help-contact-confirmation/style
   uses @include
me/help/help-contact-form/style
   uses @include
me/help/help-contact/style
   uses @include
me/help/help-happiness-engineers/style
   uses @include
me/help/style
   uses @include
me/pending-payments/style
   uses @include
me/purchases/cancel-purchase/style
   uses @include
me/purchases/components/loading-placeholder/style
   uses @include
me/purchases/manage-purchase/plan-details/style
   uses @include
me/purchases/manage-purchase/style
   uses @include
me/purchases/purchase-item/style
   uses @include
me/purchases/remove-purchase/style
   uses @include
me/security-account-recovery/style
   uses @include
my-sites/activity/activity-log-example/style
   uses @include
my-sites/activity/activity-log-switch/style
   uses @include
my-sites/checklist/wpcom-checklist/checklist-banner/style
   uses @include
my-sites/checkout/checkout-thank-you/style
   uses @include
my-sites/checkout/checkout/subscription-text
   uses @include
my-sites/checkout/concierge-session-nudge/style
   uses @include
my-sites/checkout/gsuite-nudge/style
   uses @include
my-sites/domains/domain-management/transfer/transfer-to-other-user/style
   uses @include
my-sites/email/email-forwarding/style
   uses @include
my-sites/feature-comparison/style
   uses @include
my-sites/guided-transfer/style
   uses @include
my-sites/people/people-invites/style
   uses @include
my-sites/people/people-list-item/style
   uses @include
my-sites/people/people-list-section-header/style
   uses @include
my-sites/people/people-profile/style
   uses @include
my-sites/plan-price/style
   uses @include
my-sites/plans-features-main/style
   uses @include
my-sites/plans/current-plan/current-plan-thank-you-card/style
   uses @include
my-sites/plans/current-plan/style
   uses @include
my-sites/post-type-filter/style
   uses @include
my-sites/post/post-image/style
   uses @include
my-sites/stats/annual-site-stats/style
   uses @include
my-sites/stats/post-trends/style
   uses @include
my-sites/stats/stats-tabs/style
   uses @include
my-sites/stats/stats-views/style
   uses @include
my-sites/theme/theme-download-card/style
   uses @include
my-sites/themes/current-theme/style
   uses @include
my-sites/themes/theme-upload/style
   uses @include
my-sites/themes/themes-banner/style
   uses @include
my-sites/themes/themes-selection-header/style
   uses @include
notices/style
   uses @include

SCORE: 1
components/forms/form-text-input-with-action/style
   uses @extend

SCORE: 0
blocks/daily-post-button/style
blocks/disconnect-jetpack/style
blocks/dismissible-card/style
blocks/location-search/style
blocks/plan-storage/style
blocks/plan-thank-you-card/style
blocks/upload-drop-zone/style
components/button-group/style
components/community-translator/style
components/diff-viewer/style
components/domains/example-domain-browser/style
components/domains/register-domain-step/style
components/domains/registrant-extra-info/style
components/email-verification/style
components/emojify/style
components/forms/form-button/style
components/forms/form-fieldset/style
components/forms/form-input-validation/style
components/forms/form-label/style
components/forms/form-legend/style
components/forms/form-password-input/style
components/forms/form-radio-with-thumbnail/style
components/forms/form-radios-bar/style
components/forms/form-section-heading/style
components/forms/form-select/style
components/forms/form-setting-explanation/style
components/forms/form-tel-input/style
components/forms/form-text-input/style
components/forms/form-toggle/style
components/forms/range/style
components/forms/sortable-list/style
components/gauge/style
components/header-button/style
components/hr-with-text/style
components/image/style
components/infinite-list/style
components/list-end/style
components/locale-suggestions/style
components/main/style
components/marked-lines/style
components/mini-site-preview/style
components/payment-logo/style
components/plans/plan-icon/style
components/plans/plans-skip-button/style
components/select-dropdown/style
components/signup-site-title/style
components/sites-popover/style
components/social-icons/style
components/stat-update-indicator/style
components/text-diff/style
components/themes-list/style
components/tooltip/style
jetpack-onboarding/style
layout/community-translator/style
lib/accept/style
lib/directly/style
me/account/style
me/application-password-item/style
me/application-passwords/style
me/help/help-contact-closed/style
me/help/help-unverified-warning/style
me/privacy/style
me/purchases/confirm-cancel-domain/style
me/security/style
me/two-step/style
my-sites/checklist/style
my-sites/checklist/wpcom-checklist/checklist-navigation/style
my-sites/domain-tip/style
my-sites/domains/domain-management/components/designated-agent-notice/style
my-sites/domains/domain-management/components/inbound-transfer-verification/style
my-sites/draft/style
my-sites/earn/style
my-sites/no-results/style
my-sites/pages/page-card-info/style
my-sites/people/delete-user/style
my-sites/people/edit-team-member-form/style
my-sites/people/invite-people/style
my-sites/people/people-invite-details/style
my-sites/people/people-notices/style
my-sites/plan-compare-card/style
my-sites/plan-interval-discount/style
my-sites/plans/style
my-sites/post-relative-time-status/style
my-sites/post-selector/style
my-sites/post-type-list/post-action-counts/style
my-sites/post-type-list/post-actions-ellipsis-menu/style
my-sites/post-type-list/post-type-post-author/style
my-sites/post-type-list/post-type-site-info/style
my-sites/post-type-list/style
my-sites/sites/style
my-sites/stats/all-time/style
my-sites/stats/stats-date-picker/style
my-sites/stats/stats-page-placeholder/style
my-sites/stats/stats-period-navigation/style
my-sites/stats/stats-post-likes/style
my-sites/stats/stats-post-summary/style

**Comment by blowery** [comment]
@belcherj I rejiggered how the scoring works (extends and include really don't add that much complexity, get fancier about selector root violations

**Comment by blowery** [comment]
> It's funny that the gnarly Sidebar styles get score 1 or 2 🙂 The AI apparently can't see the insidious complexity there.

Any thoughts on how to capture that complexity?

**Comment by jsnajdr** [comment]
> Any thoughts on how to capture that complexity?

If I remember correctly, it was related to how stylesheets interact with each other. Ambiguous specificity where rule order matters.

Sidebar items are `<a>` links that get a lot of their styling from `assets/stylesheets/_main.scss`. That makes it hard to isolate Sidebar with minimal self-contained CSS styles.

**Pull request merged**